File: leetcode/pjkfckr/Unique_Binary_Search_Trees.py


n = 3

def solution(n: int):
    dp = [0] * (n + 1)
    dp[0] = dp[1] = 1

    """
    각 숫자에 대해, 
    왼쪽 서브트리의 가능한 구조 수와 
    오른쪽 서브트리의 가능한 구조 수를 곱하여 전체 구조 수를 계산합니다.
    
    """
    for i in range(2, n + 1):
        for j in range(1, i + 1):
            dp[i] += dp[j - 1] * dp[i - j]

    return dp[n]

File: leetcode/0chnxxx/Unique_Binary_Search_Trees.kt
/**
 * Given an integer n, return the number of structurally unique BST's (binary search trees) which has exactly n nodes of unique values from 1 to n.
 *
 * Constraints:
 * 1 <= n <= 19
 */

fun main() {
    val n = 3

    val solution = Solution().numTrees(n)

    println(solution)
}

class Solution {
    // 시간 복잡도 : O(N^2)
    // 공간 복잡도 : O(N)
    fun numTrees(n: Int): Int {
        val dp = IntArray(n + 1)

        dp[0] = 1
        dp[1] = 1

        for (i in 2..n) {
            for (j in 0 until i) {
                dp[i] += dp[j] * dp[i - j - 1]
            }
        }

        return dp[n]
    }
}
